source/simulator/nullspace.cc

```diff
@@ -81,196 +91,296 @@ namespace aspect
+    tmp_distributed_stokes.block(introspection.block_indices.velocities).add(-1.0,correction);
 
-  template <>
-  void
-  Simulator<3>::remove_net_angular_momentum( LinearAlgebra::BlockVector &relevant_dst,
-                                             LinearAlgebra::BlockVector &tmp_distributed_stokes )
-  {
-    AssertThrow(false, ExcNotImplemented());
+    //copy into the locally relevant vector
+    relevant_dst.block(introspection.block_indices.velocities) = 
+            tmp_distributed_stokes.block(introspection.block_indices.velocities);
   }
 
-
-  template <>
-  void
-  Simulator<2>::remove_net_angular_momentum( LinearAlgebra::BlockVector &relevant_dst,
-                                             LinearAlgebra::BlockVector &tmp_distributed_stokes )
+  template <int dim>
+  void Simulator<dim>::remove_net_angular_momentum( const bool use_constant_density, 
+                                                    LinearAlgebra::BlockVector &relevant_dst,
+                                                    LinearAlgebra::BlockVector &tmp_distributed_stokes)
   {
     Assert(introspection.block_indices.velocities != introspection.block_indices.pressure,
            ExcNotImplemented());
 
     // compute and remove angular momentum from velocity field, by computing
-    // int rho V \cdot r_orth = omega  * int rho x^2
-    const unsigned int dim=2;
-
+    // \int \rho u \cdot r_orth = \omega  * \int \rho x^2    ( 2 dimensions)
+    // \int \rho r \times u =  I^{-1} \cdot \omega  (3 dimensions)
+    
     QGauss<dim> quadrature(parameters.stokes_velocity_degree+1);
     FEValues<dim> fe(mapping, finite_element, quadrature,
                      UpdateFlags(update_quadrature_points | update_JxW_values | update_values));
 
-    DoFHandler<dim>::active_cell_iterator cell;
-    std::vector<Point<dim> > q_points(quadrature.size());
-    std::vector<Vector<double> > fe_vals(quadrature.size(), Vector<double>(finite_element.n_components()));
+    typename DoFHandler<dim>::active_cell_iterator cell;
+
+    // moment of inertia and angular momentum for 3D
+    SymmetricTensor<2,dim> local_moment_of_inertia;
+    Tensor<1,dim> local_angular_momentum;
 
     // analogues to the moment of inertia and angular momentum for 2D
-    double local_scalar_moment = 0;
-    double local_scalar_angular_momentum = 0;
+    double local_scalar_moment = 0.0;
+    double local_scalar_angular_momentum = 0.0;
 
     //loop over all local cells
     for (cell = dof_handler.begin_active(); cell != dof_handler.end(); ++cell)
       if (cell->is_locally_owned())
         {
           fe.reinit (cell);
-          q_points = fe.get_quadrature_points();
-          fe.get_function_values(relevant_dst, fe_vals);
+          const std::vector<Point<dim> > &q_points = fe.get_quadrature_points();
+
+          //Get the velocity at each quadrature point
+          std::vector< Tensor<1,dim> > velocities( q_points.size() );
```

Comment:
Move the declaration of this variable out of the loop over all cells so that memory has to be allocated only once, not once per cell.